jaxopt/_src/nmf.py

```diff
@@ -0,0 +1,458 @@
+      b = osqp_dual / (jnp.maximum(jnp.maximum(inf_norm(osqp_Px), inf_norm(osqp_q)), inf_norm(U.T)) + self.eps)
+      rho_osqp = ((a / b)**0.5) * rho
+      return jnp.clip(rho_osqp, self.rho_min, self.rho_max)
+
+    raise ValueError(f"Unrecognized option {self.rho_policy} for NNLS.rho_policy")
+
+  def init_state(self, init_params,
+                 params_obj: Tuple[jnp.array, jnp.array],
+                 params_eq: None = None,
+                 params_ineq: None = None):
+    Y, W = params_obj
+    H = init_params.primal
+    U = init_params.dual_ineq
+    H_bar = init_params.primal.T
+    rho = self._compute_rho(Y, W, H, H_bar, U, 0.1)
+    state = NNLSState(
+        iter_num=jnp.asarray(0, dtype=jnp.int32),
+        error=jnp.asarray(jnp.inf),
+        primal_residuals=jnp.asarray(jnp.inf),
+        dual_residuals=jnp.asarray(jnp.inf),
+        rho=rho,
+        H_bar=H_bar,
+    )
+    return state
+
+  def _compute_H_bar(self, H, G, F, U, rho, H_bar):
+    # solution to argmin_H \|H - H_bar + U\|_2 + r(H)
+    # G is PSD so G + rho * I is PSD => Conjugate Gradient can be used.
+    def matmul(vec):
+      return jnp.dot(G, vec) + rho * vec
+    right_member = F + rho * (H + U).T
+    H_bar, _ = jax.scipy.sparse.linalg.cg(matmul, right_member,
+                                          x0=H_bar, tol=self.cg_tol)
+    return H_bar
+
+  def update(self, params, state, params_obj, params_eq, params_ineq):
+    """Update state of the NNLS.
+
+    n: number of rows
+    m: number of columns
+    k: rank of low rank factorization
+
+    Args:
+      params: KKTSolution tuple, with params.primal = H and H of shape (m, k)
+      state: NNLSState object.
+      params_obj: pair (Y, W), Y of shape (n, m) and W of shape (n, k)
+      params_eq: None, present for signature purposes.
+      params_ineq: None, present for signature purposes.
+
+    Returns:
+      pair params, 
+    """
+    Y, W = params_obj
+    F = W.T @ Y
+    G = W.T @ W  # PSD matrix.
+    H, U = params.primal, params.dual_ineq
+
+    # ADMM first inner problem.
+    H_bar = self._compute_H_bar(H, G, F, U, state.rho, state.H_bar)
+
+    # ADMM second inner problem.
+    H = jax.nn.relu(H_bar.T - U)
+
+    # Gradient ascent step on dual variables.
+    U = U + (H - H_bar.T)
+
+    primal_residuals, dual_residuals = self.compute_residuals(params.primal, H, H_bar, U)
+    error = jnp.maximum(primal_residuals, dual_residuals)
+
+    rho = self._compute_rho(Y, W, H, H_bar, U, state.rho)
+
+    next_params = base.KKTSolution(primal=H, dual_eq=None, dual_ineq=U)
+
+    next_state = NNLSState(
+        iter_num=state.iter_num+1,
+        error=error,
+        primal_residuals=primal_residuals,
+        dual_residuals=dual_residuals,
+        rho=rho,
+        H_bar=H_bar,
+    )
+    return base.OptStep(next_params, next_state)
+
+  def run(self,
+          init_params: Optional[base.KKTSolution],
+          params_obj: Tuple[jnp.array, jnp.array],
+          params_eq: None = None,
```

Comment:
Why do you keep `params_eq` and `params_ineq`? I think signature compatibility is only useful if one class can be used as a drop-in replacement for another.

Reply:
They are present by default in the `optimality_fun` returned by `make_kkt_optimality_fun` so we need to keep thos for signature compatibility, unless we modify `make_kkt_optimality_fun`.